keps/sig-apps/20181220-statefulset-volume-expansion.md

```diff
@@ -0,0 +1,95 @@
+
+During the StatefulSet update process, the StatefulSet controller will detect an update to a `volumeClaimTemplate` by comparing the updated and current revision of the StatefulSet. This requires the `VolumeClaimTemplates` component of the StatefulSet to be recorded in the StatefulSet's `ControllerRevision` object. 
+
+While updating a pod, the StatefulSet controller will update a referenced persistent volume claim object if its storage request in the associated `volumeClaimTemplate` has been increased.
+
+The Statefulset controller will delete and recreate the pod after the `FileSystemResizePending` condition is True on all updated persistent volume claims it references. However, if the `ExpandInUsePersistentVolumes` feature is enabled and a volume supports online expansion the controller will not wait for the `FileSystemResizePending` condition. Instead, it will wait for the entire expansion to complete on these volumes before proceeding thereby, ensuring that the pod is not deleted while a file system resize is taking place.
+
+Note that the Statefulset controller does not support in place updates for any pod modifications. Therefore, even if all referenced volumes support online expansion the associated pods will still need to be deleted and recreated before completing the update. Support for in place updates may be introduced at a later point in time. Additionally, the best way too handle volume types that can only be expanded offline needs to be determined. These volume types require the pod to be offline before volume expansion can be initiated.
+
+The functionality provided by this enhancement will be gated by the `StatefulSetVolumeExpansion` feature gate.
+
+### Risks and Mitigations
+
+If the `ExpandInUsePersistentVolumes` alpha feature in Kubernetes v1.11 is not enabled when expanding a volume, then pods referencing the volume must be deleted and recreated for file system expansion to occur after the `FileSystemResizePending` condition is True on the persistent volume claim. If the StatefulSet is configured with the `RollingUpdate` update strategy, then the StatefulSet controller would need to wait for the `FileSystemResizePending` condition to be satisfied on each persistent volume claim referenced by the pod it is updating before restarting the pod. This could noticeably increase the update time of a StatefulSet with many replicas. 
+
+A potential mitigation would be the eventual adoption of the `ExpandInUsePersistentVolumes` feature, which enables file system expansion on in-use volumes thus, making it unecessary to delete and recreate any refrencing pods during volume expansion.
```

Comment:
This is not really a mitigation if we are always going to restart the pods anyways right? Do we need tor remove this section?

Reply:
Once in place updates become possible, it would be a mitigation. I'm fine with leaving it in or taking it out.

Reply:
I think we should take this out, because design above proposes we are *always* going to restart the pods. So even if online resizing was default, now will wait for entire operation to finish before delete and re-creating the pod.